### ISSUE TITLE
Memory leak with VImage::text() ,  vips-8.8.4

### ISSUE THREAD

**Issue by uint128** [libvips/libvips#1508]
Firstly , thanks for the amazing library!
I am using vips-8.8.4.
When I using VImage::text() to draw some cjk unicode text to new image,  the resident memory by the process increases rapidly, and never decrease.
For the example in attach, the process used about 5G memory.
Another strange point is that  the resident memory by the process  can go much bigger than the total used memory show by top command, as showing in the screen shot below:

![2019-12-21 12-36-12 screen shot](https://user-images.githubusercontent.com/2351821/71303062-b2f7e300-23ee-11ea-9268-2721c00c358d.png)

```c++
#include <iostream>
#include <string>
#include <fstream>
#include <list>
#include <boost/algorithm/string/trim.hpp>
#include <boost/lexical_cast.hpp>
#include <vips/vips8>
#include <vips/VImage8.h>

using namespace std;
using namespace boost;
using namespace vips;

int main(int argc, char* argv[] ){
  if (VIPS_INIT ("vips_test1")){ 
    vips_error_exit ("can not init vips");
	}


	string font_str = "AR PL UMing CN " + boost::lexical_cast<string>(20) + "px";
	ifstream ifs("text_to_draw.txt");
	list<string> str_ls;
	string line;
	while(getline(ifs, line)){
		trim(line);
		if( line.empty() ){
			continue;
		}
		str_ls.push_back(line);
	}

	for(int i=0; i < 10 ; i++ ){
		for(auto& line : str_ls ){
			char* escape = g_markup_escape_text(line.c_str(), -1);
			std::string escape_str(escape);
			g_free(escape);
			int width = rand()%200 + 100;
			int height = rand()%200 + 100;
			auto opt = VImage::option ()->
				set ("width", width)->
				set ("height", height)->
				set ("font", font_str.c_str());
			auto tmp_p = escape_str.c_str();
			VImage t_img = VImage::text(tmp_p, opt);
		}
	}

	// to watch the memory
	string tmp;
	cin >> tmp;
	return 0;
}
```

[text_to_draw.txt](https://github.com/libvips/libvips/files/3991063/text_to_draw.txt)

**Comment by jcupitt** [comment]
Wow, I see terrible memory behaviour here too.

Thanks, I'll investigate.

**Comment by jcupitt** [comment]
I see the problem with this simpler program:

```c++
/* compile with 
 *
 * g++ -g -Wall text.cpp `pkg-config vips-cpp --cflags --libs`
 */

#include <iostream>
#include <string>
#include <vips/vips8>

using namespace std;
using namespace vips; 

int 
main (int argc, char* argv[]) 
{       
        if (VIPS_INIT ("vips_test1"))
                vips_error_exit ("can not init vips");

        // turn off the operation cache, or we'll just run the text once
        vips_cache_set_max (0);

        //string txt = "积蓄了眼睛的疲劳使之恢复精神! ・舒适的蒸汽持续约12分钟";
        string txt = "Hello, world!";

        for (int i = 0; i < 1000; i++) {
                cout << "iteration " << i << endl;
                        
                auto tmp_p = txt.c_str();
                VImage t_img = VImage::text(tmp_p, VImage::option ()->
                        set ("width", 500)->
                        set ("height", 500));
        }

        // pause to watch the memory
        string tmp;
        cin >> tmp;
        return 0;
}
```

It gets to 2gb of memory by about 500 iterations with the English text (and by 50 iterations with the Chinese text).

**Comment by uint128** [comment]
I am currently  located to the stack:
libvips/create/text.c:369
		if( vips_text_autofit( text ) )
libvips/create/text.c:299
		if( vips_text_get_extents( text, &extents ) )
libvips/create/text.c:190
	pango_layout_get_pixel_extents( text->layout, 
it seems like the function vips_text_get_extents( ) using remarkable memory,  and it is in a long running loop. 
inside vips_text_get_extents( )  there is a calling to pango_layout_get_pixel_extents( ).

**Comment by jcupitt** [comment]
Yes, it looks like a Pango leak, or perhaps I'm misusing `pango_layout_get_pixel_extents( )`.

I'm planning to build pango myself and look in there.

**Comment by jcupitt** [comment]
Huh it was really simple, there was just a missing unref. Your example runs in a steady and small amount of memory for me now.

Could you test with this patch?

**Comment by uint128** [comment]
I have tested with the patch. The example and my other code works fine.

**Comment by jcupitt** [comment]
That's great, thank you for reporting and testing this issue.

This fix will be in 8.9.0, due in a few days.
